File: projects/18-ledger-payments-core/src/metrics.rs
//! Prometheus metrics for the observability checklist.
//!
//! The [`metrics`] facade writes to a process-global recorder, so the counter/gauge
//! call sites (in [`routes`](crate::routes), [`isolation`](crate::isolation), and
//! [`webhooks`](crate::webhooks)) stay decoupled from this wiring — they just name a
//! metric. Until [`install`] sets a recorder the macros are no-ops, which is exactly
//! what tests want (no setup, no panics).
//!
//! [`install`] returns a [`PrometheusHandle`]; render it from `/metrics` (see
//! [`crate::routes::metrics_router`]). The series the SPEC grades:
//! - [`TRANSFERS_TOTAL`] — transfers posted, labelled by `result`.
//! - [`SERIALIZATION_RETRIES_TOTAL`] — 40001 retries (V2 contention).
//! - [`IDEMPOTENCY_LOOKUPS_TOTAL`] — labelled `outcome = hit|miss`; `hit / sum` is the ratio.
//! - [`WEBHOOK_DELIVERIES_TOTAL`] — labelled `state = delivered|failed|dead`.
//! - [`WEBHOOK_OUTBOX_LAG`] — age of the oldest pending outbox event (a gauge).
//!
//! Wiring the *call sites* is the observability horizontal item — this module just
//! makes `/metrics` render and single-sources the metric names.

use metrics_exporter_prometheus::{PrometheusBuilder, PrometheusHandle};

/// Transfers posted, labelled `result = ok|rejected|conflict|exhausted`.
pub const TRANSFERS_TOTAL: &str = "ledger_transfers_total";

/// SERIALIZABLE conflict (SQLSTATE 40001) retries — how hard V2 is fighting.
pub const SERIALIZATION_RETRIES_TOTAL: &str = "ledger_serialization_retries_total";

/// Idempotency lookups, labelled `outcome = hit|miss`. Hit ratio = `hit / (hit+miss)`.
pub const IDEMPOTENCY_LOOKUPS_TOTAL: &str = "ledger_idempotency_lookups_total";

/// Webhook delivery attempts resolved, labelled `state = delivered|failed|dead`.
pub const WEBHOOK_DELIVERIES_TOTAL: &str = "ledger_webhook_deliveries_total";

/// Outbox lag: seconds the oldest still-pending event has been waiting (a gauge).
pub const WEBHOOK_OUTBOX_LAG: &str = "ledger_webhook_outbox_lag_seconds";

/// Install the process-global Prometheus recorder and return a handle used to render
/// the registry for `/metrics`. Call once, from `main`, after telemetry init. Panics
/// if a recorder is already installed (calling it twice is a bug).
pub fn install() -> PrometheusHandle {
    register_descriptions();
    PrometheusBuilder::new()
        .install_recorder()
        .expect("install prometheus recorder")
}

/// Register HELP metadata for the graded metrics. Naming each constant here also
/// gives rustc a direct use of it (proc-macro call sites alone don't always satisfy
/// `dead_code`), so the scaffold stays warning-quiet before the call sites exist.
fn register_descriptions() {
    let transfers = TRANSFERS_TOTAL;
    let retries = SERIALIZATION_RETRIES_TOTAL;
    let idem = IDEMPOTENCY_LOOKUPS_TOTAL;
    let webhooks = WEBHOOK_DELIVERIES_TOTAL;
    let lag = WEBHOOK_OUTBOX_LAG;
    metrics::describe_counter!(transfers, "Transfers posted, labelled by result");
    metrics::describe_counter!(retries, "SERIALIZABLE 40001 retries (V2 contention)");
    metrics::describe_counter!(idem, "Idempotency lookups, labelled hit|miss");
    metrics::describe_counter!(
        webhooks,
        "Webhook deliveries, labelled delivered|failed|dead"
    );
    metrics::describe_gauge!(
        lag,
        "Age of the oldest pending webhook outbox event, seconds"
    );
}

File: projects/18-ledger-payments-core/src/ledger_payments_core/metrics.py
"""Prometheus metrics for the observability checklist.

`prometheus_client` collectors register themselves into the default registry at
import time, and `common_telemetry.metrics_routes()` renders that same registry at
`/metrics` — so declaring a metric here is all the wiring there is. That replaces the
Rust side's install-a-global-recorder step entirely.

The series the SPEC grades:

* `TRANSFERS_TOTAL` — transfers, labelled `result = ok|rejected|conflict|exhausted`;
  `rate()` over it is transfers/sec.
* `SERIALIZATION_RETRIES_TOTAL` — conflict retries: how hard V2 is fighting.
* `IDEMPOTENCY_LOOKUPS_TOTAL` — labelled `outcome = hit|miss`. The hit ratio is
  `hit / (hit + miss)`, computed at query time.
* `WEBHOOK_DELIVERIES_TOTAL` — labelled `state = delivered|failed|dead`.
* `WEBHOOK_OUTBOX_LAG_SECONDS` — age of the oldest still-pending outbox event.

Plus one the boss fight leans on: `TRANSFER_DURATION_SECONDS`, the server-side view
of the p99 ≤ 25 ms target.

Note what is *not* here: an `idempotency_hit_ratio` gauge. Counters per outcome,
never a pre-computed ratio — a ratio can't be aggregated across replicas or
re-windowed after the fact; two counters can, and Prometheus divides.

Wiring the *call sites* (the transfer path, the idempotency store, the dispatcher)
is the Observability checklist item. This module declares the series and makes
`/metrics` render them.
"""

from __future__ import annotations

from prometheus_client import Counter, Gauge, Histogram

__all__ = [
    "IDEMPOTENCY_LOOKUPS_TOTAL",
    "SERIALIZATION_RETRIES_TOTAL",
    "TRANSFERS_TOTAL",
    "TRANSFER_DURATION_SECONDS",
    "WEBHOOK_DELIVERIES_TOTAL",
    "WEBHOOK_OUTBOX_LAG_SECONDS",
    "init",
]

TRANSFER_RESULTS = ("ok", "rejected", "conflict", "exhausted")
IDEMPOTENCY_OUTCOMES = ("hit", "miss")
DELIVERY_STATES = ("delivered", "failed", "dead")

# ---- Counters (rates) ------------------------------------------------------

TRANSFERS_TOTAL = Counter(
    "ledger_transfers_total",
    "Transfers attempted, labelled by result",
    ["result"],
)

SERIALIZATION_RETRIES_TOTAL = Counter(
    "ledger_serialization_retries_total",
    "Transfers retried after a serialization conflict (V2 contention)",
)

IDEMPOTENCY_LOOKUPS_TOTAL = Counter(
    "ledger_idempotency_lookups_total",
    "Idempotency key lookups; hit ratio = hit / (hit + miss)",
    ["outcome"],
)

WEBHOOK_DELIVERIES_TOTAL = Counter(
    "ledger_webhook_deliveries_total",
    "Webhook delivery attempts resolved, labelled delivered|failed|dead",
    ["state"],
)

# ---- Gauges (current state) ------------------------------------------------

WEBHOOK_OUTBOX_LAG_SECONDS = Gauge(
    "ledger_webhook_outbox_lag_seconds",
    "Age of the oldest pending webhook outbox event, in seconds",
)
"""Set by the dispatcher each round with `set`, not `set_function`: the value lives
in Postgres, and a scrape-time callback can't await a query."""

# ---- Histograms (distributions) --------------------------------------------

TRANSFER_DURATION_SECONDS = Histogram(
    "ledger_transfer_duration_seconds",
    "Server-side time to settle one transfer, retries included",
    # The boss fight grades p99 ≤ 25ms, so the buckets need resolution *around*
    # 25ms; a histogram whose nearest edges are 10ms and 100ms cannot tell you
    # whether you passed.
    buckets=(0.001, 0.0025, 0.005, 0.01, 0.015, 0.02, 0.025, 0.03, 0.05, 0.1, 0.25, 1.0),
)


def init() -> None:
    """Pre-create every known label set, so each series exists at zero from the
    first scrape.

    A labelled collector renders *nothing* until some label combination has been
    used: `prometheus_client` creates a child on the first `.labels(...)` call. On a
    freshly started ledger that means `ledger_webhook_deliveries_total{state="dead"}`
    is simply absent — so an alert on its `rate()` can't fire, and you find out about
    the DLQ during the incident rather than before it. Idempotent: `.labels()`
    returns the existing child on a repeat call.
    """
    for result in TRANSFER_RESULTS:
        TRANSFERS_TOTAL.labels(result=result)
    for outcome in IDEMPOTENCY_OUTCOMES:
        IDEMPOTENCY_LOOKUPS_TOTAL.labels(outcome=outcome)
    for state in DELIVERY_STATES:
        WEBHOOK_DELIVERIES_TOTAL.labels(state=state)
